pkg/bpf/exporter.go

```diff
@@ -192,38 +193,221 @@ func (e *exporter) Detach() {
-	maxEntries := e.bpfObjects.Processes.MaxEntries()
-	total := 0
-	deleteKeys := make([]uint32, maxEntries)
-	deleteValues := make([]ProcessMetrics, maxEntries)
-	var cursor ebpf.MapBatchCursor
+func (e *exporter) Start(stopChan <-chan struct{}) error {
+	rd, err := ringbuf.NewReader(e.bpfObjects.Rb)
+	if err != nil {
+		return fmt.Errorf("failed to create ring buffer reader: %v", err)
+	}
+	defer rd.Close()
+
 	for {
-		count, err := e.bpfObjects.Processes.BatchLookupAndDelete(
-			&cursor,
-			deleteKeys,
-			deleteValues,
-			&ebpf.BatchOptions{},
-		)
-		total += count
-		if errors.Is(err, ebpf.ErrKeyNotExist) {
-			break
+		var record *ringbuf.Record
+
+		select {
+		case <-stopChan:
+			if err := rd.Close(); err != nil {
+				return fmt.Errorf("closing ring buffer reader: %v", err)
+			}
+			return nil
+		default:
+			var event keplerEvent
+			record = new(ringbuf.Record)
+
+			err := rd.ReadInto(record)
+			if err != nil {
+				if errors.Is(err, ringbuf.ErrClosed) {
+					return nil
+				}
+				if errors.Is(err, ringbuf.ErrFlushed) {
+					record.RawSample = record.RawSample[:0]
+				}
+				klog.Errorf("reading from reader: %s", err)
+				continue
+			}
+
+			if err := binary.Read(bytes.NewBuffer(record.RawSample), binary.LittleEndian, &event); err != nil {
+				klog.Errorf("parsing ringbuf event: %s", err)
+				continue
+			}
+
+			// Process the event
+			e.handleEvent(event)
 		}
-		if err != nil {
-			return nil, fmt.Errorf("failed to batch lookup and delete: %v", err)
+	}
+}
+
+type bpfMetrics struct {
+	CGroupID        uint64
+	CPUCyles        PerCPUCounter
+	CPUInstructions PerCPUCounter
+	CacheMiss       PerCPUCounter
+	CPUTime         PerCPUCounter
+	TxIRQ           uint64
+	RxIRQ           uint64
+	BlockIRQ        uint64
+	PageCacheHit    uint64
+}
+
+func (p *bpfMetrics) Reset() {
+	p.CPUCyles.Reset()
+	p.CPUInstructions.Reset()
+	p.CacheMiss.Reset()
+	p.CPUTime.Reset()
+	p.TxIRQ = 0
+	p.RxIRQ = 0
+	p.BlockIRQ = 0
+	p.PageCacheHit = 0
+}
+
+func newBpfMetrics() *bpfMetrics {
+	return &bpfMetrics{
+		CPUCyles:        NewPerCPUCounter(),
+		CPUInstructions: NewPerCPUCounter(),
+		CacheMiss:       NewPerCPUCounter(),
+		CPUTime:         NewPerCPUCounter(),
+	}
+}
+
+type PerCPUCounter struct {
+	Values map[uint64]uint64
+	Total  uint64
+}
+
+func NewPerCPUCounter() PerCPUCounter {
+	return PerCPUCounter{
+		Values: make(map[uint64]uint64),
+	}
+}
+
+func (p *PerCPUCounter) Start(cpu, taskID uint32, value uint64) error {
+	key := uint64(cpu)<<32 | uint64(taskID)
+
+	// TODO: The eBPF code would blindly overwrite the value if it already exists.
+	// We will preserve the old behavior for now, but we should consider
+	// returning an error if the value already exists.
+	p.Values[key] = value
+
+	return nil
+}
+
+func (p *PerCPUCounter) Stop(cpu, taskID uint32, value uint64) error {
+	key := uint64(cpu)<<32 | uint64(taskID)
+
+	if _, ok := p.Values[key]; !ok {
+		return nil
+	}
+
+	if value == 0 {
+		return nil
+	}
```

Comment:
```suggestion
	if value == 0 {
		return nil
	}
	
	key := uint64(cpu)<<32 | uint64(taskID)
	if _, ok := p.Values[key]; !ok {
		return nil
	}

```

Lets avoid a lookup when possible